File: web/packages/teleport/src/Integrations/Enroll/Cloud/Aws/regions.ts
/**
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { awsRegionMap } from 'teleport/services/integrations';

import { Region, RegionGroup, RegionId } from '../RegionMultiSelector/types';

const regions: { name: string; regions: RegionId[] }[] = [
  {
    name: 'North America',
    regions: [
      'us-east-1',
      'us-east-2',
      'us-west-1',
      'us-west-2',
      'ca-central-1',
    ],
  },
  {
    name: 'South America',
    regions: ['sa-east-1'],
  },
  {
    name: 'Asia Pacific',
    regions: [
      'ap-east-1',
      'ap-northeast-1',
      'ap-northeast-2',
      'ap-northeast-3',
      'ap-south-1',
      'ap-south-2',
      'ap-southeast-1',
      'ap-southeast-2',
      'ap-southeast-3',
      'ap-southeast-4',
    ],
  },
  {
    name: 'Europe',
    regions: [
      'eu-central-1',
      'eu-central-2',
      'eu-north-1',
      'eu-south-1',
      'eu-south-2',
      'eu-west-1',
      'eu-west-2',
      'eu-west-3',
    ],
  },
  {
    name: 'Middle East',
    regions: ['me-south-1', 'me-central-1', 'il-central-1'],
  },
  {
    name: 'Africa',
    regions: ['af-south-1'],
  },
];

function createAwsRegionGroups(): readonly RegionGroup[] {
  return regions.map(({ name, regions }) => ({
    name,
    regions: regions
      .filter(id => id in awsRegionMap)
      .map(id => ({
        id,
        name: awsRegionMap[id],
      })) as Region[],
  }));
}

export const awsRegionGroups = createAwsRegionGroups();

File: lib/cloud/aws/regions/regions.go
/*
 * Teleport
 * Copyright (C) 2026  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package regions

import (
	"context"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/account"
	accounttypes "github.com/aws/aws-sdk-go-v2/service/account/types"
	"github.com/gravitational/trace"

	libcloudaws "github.com/gravitational/teleport/lib/cloud/aws"
	"github.com/gravitational/teleport/lib/cloud/awsconfig"
)

// ListerGetter gets a client that is capable of listing AWS regions.
type ListerGetter func(ctx context.Context, opts ...awsconfig.OptionsFn) (account.ListRegionsAPIClient, error)

// ListEnabledRegions returns every region enabled for the caller's AWS
// account, using the account:ListRegions API.
func ListEnabledRegions(ctx context.Context, listerGetter ListerGetter, opts ...awsconfig.OptionsFn) ([]string, error) {
	regionsListerClient, err := listerGetter(ctx, opts...)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	paginator := account.NewListRegionsPaginator(regionsListerClient, &account.ListRegionsInput{
		RegionOptStatusContains: []accounttypes.RegionOptStatus{
			accounttypes.RegionOptStatusEnabled,
			accounttypes.RegionOptStatusEnabledByDefault,
		},
	})

	var enabledRegions []string
	for paginator.HasMorePages() {
		page, err := paginator.NextPage(ctx)
		if err != nil {
			convertedErr := libcloudaws.ConvertRequestFailureError(err)
			if trace.IsAccessDenied(convertedErr) {
				return nil, trace.BadParameter("Missing account:ListRegions permission in IAM Role, which is required to iterate over all regions. " +
					"Add this permission to the IAM Role, or enumerate the regions explicitly.")
			}
			return nil, convertedErr
		}

		for _, region := range page.Regions {
			enabledRegions = append(enabledRegions, aws.ToString(region.RegionName))
		}
	}

	return enabledRegions, nil
}
